File: smalivm/src/main/java/org/cf/smalivm/opcode/MonitorEnterOp.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.context.ExecutionNode;
import org.cf.smalivm.context.MethodState;
import org.jf.dexlib2.builder.MethodLocation;

public class MonitorEnterOp extends MethodStateOp {

    private final int destRegister;

    MonitorEnterOp(MethodLocation location, MethodLocation child, int destRegister) {
        super(location, child);
        this.destRegister = destRegister;
    }

    @Override
    public void execute(ExecutionNode node, MethodState mState) {
    }

    @Override
    public String toString() {
        return getName() + " r" + destRegister;
    }
}

File: smalivm/src/main/java/org/cf/smalivm/opcode/MonitorEnterOp.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm2.ExecutionNode
import org.cf.smalivm2.UnresolvedChild
import org.jf.dexlib2.builder.BuilderInstruction
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.OneRegisterInstruction

class MonitorEnterOp internal constructor(location: MethodLocation, private val destRegister: Int) :
    Op(location) {

    override val registersReadCount = 0
    override val registersAssignedCount = 0

    override fun execute(node: ExecutionNode): Array<out UnresolvedChild> {
        return finish()
    }

    override fun toString() = "$name r$destRegister"

    companion object : OpFactory {
        override fun build(
            location: MethodLocation,
            classManager: ClassManager,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val instruction = location.instruction as BuilderInstruction
            val destRegister = (instruction as OneRegisterInstruction).registerA
            return MonitorEnterOp(location, destRegister)
        }
    }
}
